pkg/controller/clusterclaim/clusterclaim_controller.go

```diff
@@ -117,20 +114,37 @@ func claimForClusterDeployment(cd *hivev1.ClusterDeployment) *types.NamespacedNa
 	}
 }
 
-func requestsForClusterDeployment(ctx context.Context, o client.Object) []reconcile.Request {
-	cd, ok := o.(*hivev1.ClusterDeployment)
-	if !ok {
-		return nil
-	}
+func requestsForClusterDeployment(ctx context.Context, cd *hivev1.ClusterDeployment) []reconcile.Request {
+	// FIXME how to handle when object is not a ClusterDeployment?
 	claim := claimForClusterDeployment(cd)
 	if claim == nil {
 		return nil
 	}
 	return []reconcile.Request{{NamespacedName: *claim}}
 }
 
-func requestsForRBACResources(c client.Client, resourceName string, logger log.FieldLogger) handler.MapFunc {
-	return func(ctx context.Context, o client.Object) []reconcile.Request {
+func requestsForRBACResourcesRole(c client.Client, resourceName string, logger log.FieldLogger) handler.TypedMapFunc[*rbacv1.Role] {
+	return func(ctx context.Context, o *rbacv1.Role) []reconcile.Request {
+		if o.GetName() != resourceName {
+			return nil
+		}
+		clusterName := o.GetNamespace()
+		cd := &hivev1.ClusterDeployment{}
+		if err := c.Get(context.Background(), client.ObjectKey{Namespace: clusterName, Name: clusterName}, cd); err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "failed to get ClusterDeployment for RBAC resource")
+			return nil
+		}
+		claim := claimForClusterDeployment(cd)
+		if claim == nil {
+			return nil
+		}
+		return []reconcile.Request{{NamespacedName: *claim}}
+	}
+}
+
+// FIXME refrain from duplicating?
```

Comment:
Do generics allow type unions?

Alternatively, you could make your internal func accept `metav1.Object`, since you only need it to respond to `GetName()` and `GetNamespace()`.